### PR TITLE
[core] Fix the file descriptor leaking when worker or node died in GCS

### LINKED ISSUE #33241: [core] Connection leak in gRPC
### What happened + What you expected to happen

When an actor is killed, the connection to that actor should be disconnected. But it turns out not.

This is very bad since the max number of actors a cluster can create is limited to rlimit of files in the system.

### Versions / Dependencies

master

### Reproduction script

Start ray cluster create actor in driver. Exit the driver check FDs opened.

```
ray start --head
```

And then `lsof -a GCS_PID | wc -l`
It shows 40.
```
import ray

ray.init()

import timeit

@ray.remote
class A:
    def f(self): return "HELLO"

a = [A.remote() for _ in range(10)]
print(">>>", len(a))
print(ray.get([x.f.remote() for x in a]))
```
And then `lsof -a GCS_PID | wc -l`

It shows 62

Check the value before and after. They should be the same.
### Issue Severity

None

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
Ray GCS pubsub right now is implemented in the way of long polling. An RPC has to be closed to free the connection timely. Otherwise, it won't be closed.

In case of worker died or node died, the cleanup in GCS pubsub is not done well. It depends on Ray pubsub's liveness check to cleanup the inactivate long poll.

This PR tried to close the RPC in case of worker or node failure.

The case is a little bit tricky for GCS client. In python worker, there are two GCS connections, one is from cpp and the other one is from python. And this means, we can't use set subscriber id  to worker id directly since cpp and py should have different subscriber id.

So a new field called sent_from is introduced to protobuf. If it's set, this field will be used as the node id and if not, it means using subscriber_id.

All logic is implemented in GCS pubsub handler. In CoreWorker <-> CoreWorker, we still depends on pubsub liveness check to close the connections. But given that the connections between core workers are not a lot, it should be OK.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/pubsub/publisher.h [hunk 1]
```diff
@@ -179,7 +179,11 @@ class SubscriberState {
         publish_batch_size_(publish_batch_size),
         last_connection_update_time_ms_(get_time_ms_()) {}
 
-  ~SubscriberState() = default;
+  ~SubscriberState() {
+    // Force a push to close the long-polling.
+    // Otherwise, there will be a connection leak.
+    PublishIfPossible(true);
```
Comment (rkooo567) [inline src/ray/pubsub/publisher.h:185]: nice catch!

**Review by rkooo567 — APPROVED** [review]
LGTM. Btw, is a reason why CheckDeadSubscribers not enough because the timeout is too long (5m) right?

**Review by rkooo567 — CHANGES_REQUESTED** [review]
(sorry let me approve after the answer of comments regarding the worker id ^)

File: python/ray/_private/worker.py [hunk 2]
```diff
@@ -2124,6 +2117,17 @@ def connect(
 
     # Notify raylet that the core worker is ready.
     worker.core_worker.notify_raylet()
+    worker_id = worker.worker_id
```
[the remark this answers is not in the document]
Comment (iycheng) [reply python/ray/_private/worker.py:2123]: I think it comes from here: https://sourcegraph.com/github.com/ray-project/ray@master/-/blob/python/ray/_raylet.pyx?L1609:9

It's from core worker, so I think it should be the same?

**Comment by iycheng** [comment]
Test failed. Rebase to master first.

File: python/ray/_private/worker.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rkooo567) [reply python/ray/_private/worker.py:2124]: I see. You are right. Maybe I was confused with the node id or sth...

**Review by rkooo567 — APPROVED** [review]

**Review by scv119 — APPROVED** [review]
Thanks for the fix!

**Comment by iycheng** [comment]
test failure not related.

**Comment by ArturNiederfahrenhorst** [comment]
@iycheng I get some error I'm unable to parse since this commit:

> 
> /Users/artur/miniforge3/envs/ray/bin/python /Applications/PyCharm.app/Contents/plugins/python/helpers/pycharm/_jb_pytest_runner.py --path /Users/artur/code/ray/python/ray/rllib/algorithms/ppo/tests/test_ppo_rl_module.py
> Testing started at 16:26 ...
> Launching pytest with arguments /Users/artur/code/ray/python/ray/rllib/algorithms/ppo/tests/test_ppo_rl_module.py --no-header --no-summary -q in /Users/artur/code/ray/python/ray/rllib/algorithms/ppo/tests
> 
> ============================= test session starts ==============================
> collecting ... collected 2 items
> 
> ../../../../python/ray/rllib/algorithms/ppo/tests/test_ppo_rl_module.py::TestPPO::test_forward_train ERROR [ 50%]2023-03-18 16:26:58,338	INFO worker.py:1559 -- Started a local Ray instance.
> 
> test setup failed
> message_descriptor = <google.protobuf.descriptor.Descriptor object at 0x17ced0b50>
> field_name = 'sender_id'
> 
>     def _GetFieldByName(message_descriptor, field_name):
>       """Returns a field descriptor by field name.
>     
>       Args:
>         message_descriptor: A Descriptor describing all fields in message.
>         field_name: The name of the field to retrieve.
>       Returns:
>         The field descriptor associated with the field name.
>       """
>       try:
> >       return message_descriptor.fields_by_name[field_name]
> E       KeyError: 'sender_id'
> 
> /Users/artur/miniforge3/envs/ray/lib/python3.10/site-packages/google/protobuf/internal/python_message.py:577: KeyError
> 
> During handling of the above exception, another exception occurred:
> 
> cls = <class 'test_ppo_rl_module.TestPPO'>
> 
>     @classmethod
>     def setUpClass(cls):
> >       ray.init()
> 
> ../../../../python/ray/rllib/algorithms/ppo/tests/test_ppo_rl_module.py:118: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> ../../../../python/ray/_private/client_mode_hook.py:105: in wrapper
>     return func(*args, **kwargs)
> ../../../../python/ray/_private/worker.py:1561: in init
>     connect(
> ../../../../python/ray/_private/worker.py:2147: in connect
>     worker.import_thread = import_thread.ImportThread(
> ../../../../python/ray/_private/import_thread.py:37: in __init__
>     self.subscriber.subscribe()
> ../../../../python/ray/_private/gcs_pubsub.py:238: in subscribe
>     req = self._subscribe_request(self._channel)
> ../../../../python/ray/_private/gcs_pubsub.py:87: in _subscribe_request
>     req = gcs_service_pb2.GcsSubscriberCommandBatchRequest(
> /Users/artur/miniforge3/envs/ray/lib/python3.10/site-packages/google/protobuf/internal/python_message.py:516: in init
>     field = _GetFieldByName(message_descriptor, field_name)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> message_descriptor = <google.protobuf.descriptor.Descriptor object at 0x17ced0b50>
> field_name = 'sender_id'
> 
>     def _GetFieldByName(message_descriptor, field_name):
>       """Returns a field descriptor by field name.
>     
>       Args:
>         message_descriptor: A Descriptor describing all fields in message.
>         field_name: The name of the field to retrieve.
>       Returns:
>         The field descriptor associated with the field name.
>       """
>       try:
>         return message_descriptor.fields_by_name[field_name]
>       except KeyError:
> >       raise ValueError('Protocol message %s has no "%s" field.' %
>                          (message_descriptor.name, field_name))
> E       ValueError: Protocol message GcsSubscriberCommandBatchRequest has no "sender_id" field.
> 
> /Users/artur/miniforge3/envs/ray/lib/python3.10/site-packages/google/protobuf/internal/python_message.py:579: ValueError
> ERROR [100%]
> test setup failed
> message_descriptor = <google.protobuf.descriptor.Descriptor object at 0x17ced0b50>
> field_name = 'sender_id'
> 
>     def _GetFieldByName(message_descriptor, field_name):
>       """Returns a field descriptor by field name.
>     
>       Args:
>         message_descriptor: A Descriptor describing all fields in message.
>         field_name: The name of the field to retrieve.
>       Returns:
>         The field descriptor associated with the field name.
>       """
>       try:
> >       return message_descriptor.fields_by_name[field_name]
> E       KeyError: 'sender_id'
> 
> /Users/artur/miniforge3/envs/ray/lib/python3.10/site-packages/google/protobuf/internal/python_message.py:577: KeyError
> 
> During handling of the above exception, another exception occurred:
> 
> cls = <class 'test_ppo_rl_module.TestPPO'>
> 
>     @classmethod
>     def setUpClass(cls):
> >       ray.init()
> 
> ../../../../python/ray/rllib/algorithms/ppo/tests/test_ppo_rl_module.py:118: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> ../../../../python/ray/_private/client_mode_hook.py:105: in wrapper
>     return func(*args, **kwargs)
> ../../../../python/ray/_private/worker.py:1561: in init
>     connect(
> ../../../../python/ray/_private/worker.py:2147: in connect
>     worker.import_thread = import_thread.ImportThread(
> ../../../../python/ray/_private/import_thread.py:37: in __init__
>     self.subscriber.subscribe()
> ../../../../python/ray/_private/gcs_pubsub.py:238: in subscribe
>     req = self._subscribe_request(self._channel)
> ../../../../python/ray/_private/gcs_pubsub.py:87: in _subscribe_request
>     req = gcs_service_pb2.GcsSubscriberCommandBatchRequest(
> /Users/artur/miniforge3/envs/ray/lib/python3.10/site-packages/google/protobuf/internal/python_message.py:516: in init
>     field = _GetFieldByName(message_descriptor, field_name)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> message_descriptor = <google.protobuf.descriptor.Descriptor object at 0x17ced0b50>
> field_name = 'sender_id'
> 
>     def _GetFieldByName(message_descriptor, field_name):
>       """Returns a field descriptor by field name.
>     
>       Args:
>         message_descriptor: A Descriptor describing all fields in message.
>         field_name: The name of the field to retrieve.
>       Returns:
>         The field descriptor associated with the field name.
>       """
>       try:
>         return message_descriptor.fields_by_name[field_name]
>       except KeyError:
> >       raise ValueError('Protocol message %s has no "%s" field.' %
>                          (message_descriptor.name, field_name))
> E       ValueError: Protocol message GcsSubscriberCommandBatchRequest has no "sender_id" field.
> 
> /Users/artur/miniforge3/envs/ray/lib/python3.10/site-packages/google/protobuf/internal/python_message.py:579: ValueError
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> ../../../../python/ray/rllib/algorithms/ppo/tests/test_ppo_rl_module.py::TestPPO::test_rollouts 
> 
> ============================== 2 errors in 7.42s ===============================
> Exception ignored in atexit callback: <function shutdown at 0x289d9dc60>
> Traceback (most recent call last):
>   File "/Users/artur/code/ray/python/ray/_private/client_mode_hook.py", line 105, in wrapper
>     return func(*args, **kwargs)
>   File "/Users/artur/code/ray/python/ray/_private/worker.py", line 1624, in shutdown
>     disconnect(_exiting_interpreter)
>   File "/Users/artur/code/ray/python/ray/_private/worker.py", line 2220, in disconnect
>     worker.gcs_function_key_subscriber.close()
>   File "/Users/artur/code/ray/python/ray/_private/gcs_pubsub.py", line 288, in close
>     req = self._unsubscribe_request(channels=[self._channel])
>   File "/Users/artur/code/ray/python/ray/_private/gcs_pubsub.py", line 98, in _unsubscribe_request
>     req = gcs_service_pb2.GcsSubscriberCommandBatchRequest(
>   File "/Users/artur/miniforge3/envs/ray/lib/python3.10/site-packages/google/protobuf/internal/python_message.py", line 516, in init
>     field = _GetFieldByName(message_descriptor, field_name)
>   File "/Users/artur/miniforge3/envs/ray/lib/python3.10/site-packages/google/protobuf/internal/python_message.py", line 579, in _GetFieldByName
>     raise ValueError('Protocol message %s has no "%s" field.' %
> ValueError: Protocol message GcsSubscriberCommandBatchRequest has no "sender_id" field.
> [2023-03-18 16:27:00,840 E 9324 58097] logging.cc:104: Stack trace: 
>  0   _raylet.so                          0x000000028bda1d78 _ZN3raylsERNSt3__113basic_ostreamIcNS0_11char_traitsIcEEEERKNS_10StackTraceE + 84 ray::operator<<()
> 1   _raylet.so                          0x000000028bda1f64 _ZN3ray16TerminateHandlerEv + 228 ray::TerminateHandler()
> 2   libc++abi.dylib                     0x00000001b0171eb4 _ZSt11__terminatePFvvE + 20 std::__terminate()
> 3   libc++abi.dylib                     0x00000001b0171e3c _ZSt9terminatev + 36 std::terminate()
> 4   _raylet.so                          0x000000028b724eb8 _ZN3ray4core6worker19TaskEventBufferImplD2Ev + 124 ray::core::worker::TaskEventBufferImpl::~TaskEventBufferImpl()
> 5   _raylet.so                          0x000000028b724da8 _ZN3ray4core6worker19TaskEventBufferImplD0Ev + 12 ray::core::worker::TaskEventBufferImpl::~TaskEventBufferImpl()
> 6   _raylet.so                          0x000000028b62fdcc _ZN3ray4core10CoreWorkerD2Ev + 208 ray::core::CoreWorker::~CoreWorker()
> 7   _raylet.so                          0x000000028b6e2efc _ZN3ray4core21CoreWorkerProcessImplD2Ev + 664 ray::core::CoreWorkerProcessImpl::~CoreWorkerProcessImpl()
> 8   _raylet.so                          0x000000028b6e0204 _ZN3ray4core17CoreWorkerProcess12HandleAtExitEv + 28 ray::core::CoreWorkerProcess::HandleAtExit()
> 9   libsystem_c.dylib                   0x00000001b00a3fe8 __cxa_finalize_ranges + 480 __cxa_finalize_ranges
> 10  libsystem_c.dylib                   0x00000001b00a3d6c exit + 44 exit
> 11  python3.10                          0x00000001050e5ef4 Py_FdIsInteractive + 0 Py_FdIsInteractive
> 12  python3.10                          0x00000001050ec050 _PyErr_PrintEx + 2084 _PyErr_PrintEx
> 13  python3.10                          0x00000001050ec858 _PyRun_SimpleFileObject + 1324 _PyRun_SimpleFileObject
> 14  python3.10                          0x00000001050eb7f4 _PyRun_AnyFileObject + 240 _PyRun_AnyFileObject
> 15  python3.10                          0x00000001051108ac Py_RunMain + 2340 Py_RunMain
> 16  python3.10                          0x0000000105111a68 pymain_main + 1272 pymain_main
> 17  python3.10                          0x0000000104f57dc8 main + 56 main
> 18  dyld                                0x00000001afe8fe50 start + 2544 start
> 
> 
> Process finished with exit code 134 (interrupted by signal 6: SIGABRT)

Any advice?

**Comment by rkooo567** [comment]
@ArturNiederfahrenhorst you need to do bazel build //:ray_pkg

**Pull request merged**